### PR TITLE
[Merged by Bors] - feat: fix norm num with arguments

### PR DESCRIPTION
**Pull request by alexjbest**
`norm_num` was passing the wrong syntax node to `elabSimpArgs` when elaborating, which essentially had the effect of ignoring all arguments it was passed, i.e. `norm_num [add_comm]` would not try to commute addition in the simp step.
The fix itself is very simple (though not obvious to debug!), probably using TSyntax more would help avoid such issues in future.

Due to this bug many `norm_num [blah]` became `rw [blah]; norm_num` or similar, sometimes with porting notes, sometimes not, we fix these porting notes and other regressions during the port also.

Interestingly `cancel_denoms` uses `norm_num [<- mul_assoc]` internally, so  `cancel_denoms` also got stronger with this change.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
Can you summarize in the description what was fixed?

**Comment by alexjbest** [comment]
> Can you summarize in the description what was fixed?

is that ok?

File: Mathlib/Analysis/SpecialFunctions/Integrals.lean [hunk 1]
```diff
@@ -451,10 +451,7 @@ theorem integral_one_div_of_neg (ha : a < 0) (hb : b < 0) :
 
 @[simp]
 theorem integral_exp : ∫ x in a..b, exp x = exp b - exp a := by
-  rw [integral_deriv_eq_sub']
-  · norm_num
-  · exact fun _ _ => differentiableAt_exp
-  · exact continuousOn_exp
+  rw [integral_deriv_eq_sub'] <;> norm_num continuousOn_exp
```
Comment (alexjbest) [inline Mathlib/Analysis/SpecialFunctions/Integrals.lean:454]:
```suggestion
  rw [integral_deriv_eq_sub'] <;> norm_num [continuousOn_exp]
```

File: Mathlib/NumberTheory/Bernoulli.lean [hunk 2]
```diff
@@ -113,23 +113,21 @@ theorem bernoulli'_one : bernoulli' 1 = 1 / 2 := by
...
 @[simp]
 theorem bernoulli'_four : bernoulli' 4 = -1 / 30 := by
-  have : Nat.choose 4 2 = 6 := by decide
-  -- shrug
-  rw [bernoulli'_def, sum_range_succ, sum_range_succ, sum_range_succ,
-    sum_range_succ, sum_range_zero, this]
-  norm_num
+  have : Nat.choose 4 2 = 6 := by norm_num -- shrug
+  rw [bernoulli'_def]
+  norm_num [sum_range_succ, sum_range_succ, sum_range_zero]
```
Comment (alexjbest) [inline Mathlib/NumberTheory/Bernoulli.lean:130]:
```suggestion
  norm_num [sum_range_succ, sum_range_succ, sum_range_zero, this]
```

File: Mathlib/Combinatorics/Catalan.lean [hunk 3]
```diff
@@ -142,10 +142,12 @@ theorem succ_mul_catalan_eq_centralBinom (n : ℕ) : (n + 1) * catalan n = n.cen
   (Nat.eq_mul_of_div_eq_right n.succ_dvd_centralBinom (catalan_eq_centralBinom_div n).symm).symm
 #align succ_mul_catalan_eq_central_binom succ_mul_catalan_eq_centralBinom
 
-theorem catalan_two : catalan 2 = 2 := by unfold catalan; rfl
+theorem catalan_two : catalan 2 = 2 := by
+norm_num [catalan_eq_centralBinom_div, Nat.centralBinom, Nat.choose]
```
Comment (eric-wieser) [inline Mathlib/Combinatorics/Catalan.lean:146]:
```suggestion
  norm_num [catalan_eq_centralBinom_div, Nat.centralBinom, Nat.choose]
```

File: Mathlib/Combinatorics/Catalan.lean [hunk 4]
```diff
@@ -142,10 +142,12 @@ theorem succ_mul_catalan_eq_centralBinom (n : ℕ) : (n + 1) * catalan n = n.cen
...
 #align catalan_two catalan_two
 
-theorem catalan_three : catalan 3 = 5 := by unfold catalan; rfl
```
Comment (eric-wieser) [inline Mathlib/Combinatorics/Catalan.lean:148]: Why not keep the `rfl` proof?

Comment (alexjbest) [reply Mathlib/Combinatorics/Catalan.lean:148]: I assume `norm_num` will eventually become better at computing binomials than the kernel, if it isn't already.
As with most of these proofs I simply restore the mathlib3 proof unless the new one looks to have an advantage

File: Mathlib/Algebra/Ring/BooleanRing.lean [hunk 5]
```diff
@@ -56,14 +56,12 @@ variable [BooleanRing α] (a b : α)
 instance : IsIdempotent α (· * ·) :=
   ⟨BooleanRing.mul_self⟩
 
--- Porting note: simpNF linter complains. This causes lemmas in other files not to be in simpNF
```
Comment (ericrbg) [inline Mathlib/Algebra/Ring/BooleanRing.lean:59]: how could this possibly fix `simpNF`?

Comment (alexjbest) [reply Mathlib/Algebra/Ring/BooleanRing.lean:59]: The change to Mathlib/Algebra/PUnitInstances.lean is what makes the linter not complain. This is helpful to make it clear that all the `norm_num [blah]` calls that worked in mathlib 3 still work here

**Comment by semorrison** [comment]
bors merge

**Comment by digama0** [comment]
I think the diagnosis here is a bit wrong, and the fix is somewhat weird and not type-correct, although it does work. What is happening is that `elabSimpArgs` takes a syntax which is the result of matching `("[" simpArg,* "]")?`, while norm_num has a syntax of the form `(simpArgs)?` where `syntax simpArgs := "[" simpArg,* "]"`. The difference is that the former pattern has nodes which are either empty (i.e. ``node `null []``) or of the form ``node `null ["[", simpArgs, "]"]``, while the latter produces either ``node `null []`` or ``node `null [node `simpArgs ["[", simpArgs, "]"]]``.

Calling `args[0]` produces either `.missing` or ``node `simpArgs ["[", simpArgs, "]"]``, which is still not of the form expected by `elabSimpArgs` but is close enough that it has the right behavior because `elabSimpArgs` doesn't check the node kind or differentiate a null node from a missing node.

**Comment by alexjbest** [comment]
I kinda (see) thanks, do you have a better fix in mind then?

**Comment by digama0** [comment]
Not really. As you say, the solution is to use more typed syntax, but lean core is squarely at fault for using untyped syntax in this case and many others. It would also help if the `simpArgs` syntax definition was upstreamed so that all simp-like tactics use it and not just the ones defined in std/mathlib, since it makes it harder to share code between them without suspicious code like this.
